src/ray/core_worker/core_worker.cc

```diff
@@ -347,7 +347,7 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
       worker_context_(options_.worker_type, worker_id, GetProcessJobID(options_)),
       io_work_(io_service_),
       client_call_manager_(new rpc::ClientCallManager(io_service_)),
-      periodical_runner_(io_service_),
+      periodical_runner_(std::make_shared<PeriodicalRunner>(io_service_)),
```

Comment:
I think it is better passing `periodical_runner_` because if we just pass `io_service`, we will just reimplement what it is doing. 

Re; ownership. What can go wrong by using shared pointer here? (Asking for curiosity)

Reply:
I think Yi is referring to the ownership of the periodical runner being shared between the core worker and the pubsub subcomponent instead of keeping the core worker as the unique owner and passing a reference. Doing the latter would make more sense to me (and I think is more idiomatic C++) if the core worker is guaranteed to outlive the pubsub subcomponent, which is the case here, right?

When a class's lifetime encapsulates the lifetime of one of its members, and when that member wishes to share a resource with that class, tying that resource's lifetime to the class's lifetime (i.e. making the class the owner) and giving the member a reference to that resource can provide both an ownership model that's easier to reason about and better performance (no shared pointer reference counting).

Reply:
Discussed with @iycheng offline. I think performance is not a big deal here because it is passed and used only once. The lifetime stuff makes sense, so I will just do that (since it is not that difficult to implement). But note that this pattern is not followed by other parts (we can probably consider refactoring in the future).

Reply:
I saw a lot of places where we use shared pointer instead of unique pointer which is a bad pattern that makes the code harder to understand (ownership is vague) and sometimes bringing performance issues.

http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-unique

Reply:
@iycheng Do you think it makes sense to organize these things somewhere? (like some of cpp best practices we should follow)?

Reply:
Maybe it is good to have a doc (with the current state of art in our repo). That can also give us what parts need to be re-written easily

Reply:
Yes, that's a good idea. We can identify some bad patterns in our code and make a doc about them.